### PR TITLE
Add maxsurge option for control plane upgrade 

### PR DESCRIPTION
**Pull request by alexander-demicev**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes: #166 
Fixes: #193

**Special notes for your reviewer**:

**Checklist**:

- [ ] squashed commits into logical changes
- [ ] includes documentation
- [ ] adds unit tests
- [ ] adds or updates e2e tests

### CODE REVIEW

**Comment by alexander-demicev** [comment]
I will be working on tests in a follow-up PR, as we don't have any setup for unit testing of this controller

File: test/e2e/e2e_test.go [hunk 1]
```diff
@@ -148,6 +148,31 @@ var _ = Describe("Workload cluster creation", func() {
 				ControlPlane: result.ControlPlane,
 			}, e2eConfig.GetIntervals(specName, "wait-control-plane")...)
 
+			By("Upgrading control plane")
+			ApplyClusterTemplateAndWait(ctx, ApplyClusterTemplateAndWaitInput{
```
Comment (richardcase) [inline test/e2e/e2e_test.go:152]: I don't think this waits for the upgrade to have completed.

Comment (alexander-demicev) [reply test/e2e/e2e_test.go:152]: good point, I checked `WaitForControlPlaneToBeReady` function below, it's going to return true if rollout wasn't started immediately. I will add a new one `WaitForControlPlaneToUpgrade` that wait for all machines to machines to have updated spec

File: test/e2e/helpers.go [hunk 2]
```diff
@@ -251,6 +252,43 @@ func WaitForMachineConditions(ctx context.Context, input WaitForMachineCondition
...
+		if err := input.Lister.List(ctx, machineList); err != nil {
+			return false, fmt.Errorf("failed to list machines: %w", err)
+		}
+
+		if len(machineList.Items) != int(totallMachineCount) { // not all replicas are created
+			return false, nil
+		}
+
+		for _, machine := range machineList.Items {
+			if machine.Spec.Version != nil && *machine.Spec.Version != input.VersionAfterUpgrade {
```
Comment (Danil-Grigorev) [inline test/e2e/helpers.go:283]:
```suggestion
			if machine.Spec.Version == nil || *machine.Spec.Version != input.VersionAfterUpgrade {
```

File: controlplane/api/v1alpha1/rke2controlplane_types.go [hunk 3]
```diff
@@ -71,6 +72,11 @@ type RKE2ControlPlaneSpec struct {
 	// the registration type is "address". Its for scenarios where a load-balancer or VIP is used.
 	// +optional
 	RegistrationAddress string `json:"registrationAddress,omitempty"`
+
+	// The RolloutStrategy to use to replace control plane machines with new ones.
+	// +optional
+	// +kubebuilder:default={type: "RollingUpdate", rollingUpdate: {maxSurge: 1}}
```
Comment (richardcase) [inline controlplane/api/v1alpha1/rke2controlplane_types.go:78]: I was thinking that we shouldn't adopt this default like KCP to make CAPRKE2 be more GitOps friendly. 

Problem with using defaults is that it causes the original spec to differ from the one that was applied, which can then cause a GitOps implementation to think there is drift.

File: controlplane/internal/controllers/rke2controlplane_controller.go [hunk 4]
```diff
@@ -464,6 +464,14 @@ func (r *RKE2ControlPlaneReconciler) reconcileNormal(
 		return result, err
 	}
 
+	// Check if there any machines with a deletion timestamp set and requeue, we should
```
Comment (richardcase) [inline controlplane/internal/controllers/rke2controlplane_controller.go:467]:
```suggestion
	// Check if there are any machines with a deletion timestamp set and requeue, we should
```

File: controlplane/api/v1alpha1/rke2controlplane_types.go (hunk 3 above)
Comment (richardcase) [reply controlplane/api/v1alpha1/rke2controlplane_types.go:78]: Should we instead apply the default as part of the reconciliation loop?

File: controlplane/internal/controllers/rke2controlplane_controller.go [hunk 5]
```diff
@@ -742,14 +750,22 @@ func (r *RKE2ControlPlaneReconciler) upgradeControlPlane(
...
+		// We can ignore MaxUnavailable because we are enforcing health checks before we get here.
+		maxNodes := *rcp.Spec.Replicas + int32(rcp.Spec.RolloutStrategy.RollingUpdate.MaxSurge.IntValue())
+		if int32(controlPlane.Machines.Len()) < maxNodes {
+			// scaleUp ensures that we don't continue scaling up while waiting for Machines to have NodeRefs
+			return r.scaleUpControlPlane(ctx, cluster, rcp, controlPlane)
+		}
+
+		return r.scaleDownControlPlane(ctx, cluster, rcp, controlPlane, machinesRequireUpgrade)
+	default:
+		logger.Info("RolloutStrategy type is not set to RollingUpdateStrategyType, unable to determine the strategy for rolling out machines")
```
Comment (richardcase) [inline controlplane/internal/controllers/rke2controlplane_controller.go:765]: Should this be an error or warning instead?

File: test/e2e/data/infrastructure/cluster-template-docker.yaml [hunk 6]
```diff
@@ -48,7 +48,8 @@ metadata:
   name:  "${CLUSTER_NAME}-control-plane"
 spec:
   template:
-    spec: {}
+    spec:
+      customImage: kindest/node:${KIND_IMAGE_VERSION}
```
Comment (richardcase) [inline test/e2e/data/infrastructure/cluster-template-docker.yaml:52]: I'm curious, why do we need to set `customImage`?

**Review by richardcase — COMMENTED** [review]
Looking great @alexander-demicev 👍  Some small comments.

File: controlplane/api/v1alpha1/rke2controlplane_types.go (hunk 3 above)
Comment (richardcase) [reply controlplane/api/v1alpha1/rke2controlplane_types.go:78]: Or we could probably make any change around defaults in another PR, so its applied across all the APIs

Comment (alexander-demicev) [reply controlplane/api/v1alpha1/rke2controlplane_types.go:78]: Can we do it in v1alpha2?

File: test/e2e/data/infrastructure/cluster-template-docker.yaml (hunk 6 above)
Comment (alexander-demicev) [reply test/e2e/data/infrastructure/cluster-template-docker.yaml:52]: There is no kind image for all patch releases of K8S, I decided to explicitly set it here so we can use more RKE2 versions.

File: controlplane/api/v1alpha1/rke2controlplane_types.go (hunk 3 above)
Comment (richardcase) [reply controlplane/api/v1alpha1/rke2controlplane_types.go:78]: Created #196 to track changes for v1alpha2.

File: controlplane/internal/controllers/rke2controlplane_controller.go [hunk 7]
```diff
@@ -742,14 +750,23 @@ func (r *RKE2ControlPlaneReconciler) upgradeControlPlane(
 		return ctrl.Result{}, err
 	}
 
-	status := workloadCluster.ClusterStatus()
+	switch rcp.Spec.RolloutStrategy.Type {
+	case controlplanev1.RollingUpdateStrategyType:
+		// RolloutStrategy is currently defaulted and validated to be RollingUpdate
+		// We can ignore MaxUnavailable because we are enforcing health checks before we get here.
+		maxNodes := *rcp.Spec.Replicas + int32(rcp.Spec.RolloutStrategy.RollingUpdate.MaxSurge.IntValue())
+		if int32(controlPlane.Machines.Len()) < maxNodes {
```
Comment (richardcase) [inline controlplane/internal/controllers/rke2controlplane_controller.go:758]: This change will fix #193

**Review by richardcase — APPROVED** [review]

File: controlplane/internal/controllers/rke2controlplane_controller.go (hunk 7 above)
Comment (richardcase) [reply controlplane/internal/controllers/rke2controlplane_controller.go:758]: However, on testing i am seeing a panic:

```
I1115 22:17:38.254212      20 controller.go:114]  "msg"="Observed a panic in reconciler: runtime error: invalid memory address or nil pointer dereference" "RKE2ControlPlane"={"name":"test1-control-plane","namespace":"default"} "controller"="rke2controlplane" "controllerGroup"="controlplane.cluster.x-k8s.io" "controllerKind"="RKE2ControlPlane" "name"="test1-control-plane" "namespace"="default" "reconcileID"="132cfee1-80f8-4fc8-82cc-85a76bd3d997"
```

Comment (richardcase) [reply controlplane/internal/controllers/rke2controlplane_controller.go:751]: This was sorted by Danils PR

**Review by richardcase — DISMISSED** [review]

**Review by Danil-Grigorev — DISMISSED** [review]

File: controlplane/internal/controllers/rke2controlplane_controller.go [hunk 8]
```diff
@@ -742,14 +743,23 @@ func (r *RKE2ControlPlaneReconciler) upgradeControlPlane(
 		return ctrl.Result{}, err
 	}
 
-	status := workloadCluster.ClusterStatus()
+	switch rcp.Spec.RolloutStrategy.Type {
+	case controlplanev1.RollingUpdateStrategyType:
+		// RolloutStrategy is currently defaulted and validated to be RollingUpdate
+		// We can ignore MaxUnavailable because we are enforcing health checks before we get here.
```
Comment (furkatgofurov7) [inline controlplane/internal/controllers/rke2controlplane_controller.go:749]: Q: what is the use of `MaxUnavailable` in this context? AFAICT it is only MaxSurge we are introducing in this PR

File: controlplane/internal/controllers/rke2controlplane_controller.go [hunk 9]
```diff
@@ -742,14 +743,23 @@ func (r *RKE2ControlPlaneReconciler) upgradeControlPlane(
...
+		maxNodes := *rcp.Spec.Replicas + int32(rcp.Spec.RolloutStrategy.RollingUpdate.MaxSurge.IntValue())
+		if int32(controlPlane.Machines.Len()) < maxNodes {
+			// scaleUp ensures that we don't continue scaling up while waiting for Machines to have NodeRefs
```
Comment (furkatgofurov7) [inline controlplane/internal/controllers/rke2controlplane_controller.go:752]:
```suggestion
			// scaleUpControlPlane ensures that we don't continue scaling up while waiting for Machines to have NodeRefs
```

**Review by furkatgofurov7 — DISMISSED** [review]

**Review by salasberryfin — DISMISSED** [review]

File: controlplane/internal/controllers/rke2controlplane_controller.go (hunk 9 above)
Comment (richardcase) [reply controlplane/internal/controllers/rke2controlplane_controller.go]: Changed.

File: controlplane/internal/controllers/rke2controlplane_controller.go (hunk 8 above)
Comment (richardcase) [reply controlplane/internal/controllers/rke2controlplane_controller.go]: There is no use, removed the comment.

**Review by furkatgofurov7 — APPROVED** [review]

**Pull request merged**